Source/WebCore/Modules/permissions/Permissions.cpp

```diff
@@ -77,66 +82,81 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
         return;
     }
 
     auto* origin = context->securityOrigin();
     auto originData = origin ? origin->data() : SecurityOriginData { };
 
-    PermissionController::shared().query(ClientOrigin { context->topOrigin().data(), originData }, PermissionDescriptor { parameterDescriptor }, *document->page(), [this, protectedThis = Ref { *this }, parameterDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
-        auto context = m_navigator ? m_navigator->scriptExecutionContext() : nullptr;
-        if (!context || !context->globalObject())
-            return;
+    if (auto* workerGlobalScope = dynamicDowncast<WorkerGlobalScope>(context)) {
+        auto completionHandler = [originData = WTFMove(originData).isolatedCopy(), permissionDescriptor, contextIdentifier = workerGlobalScope->identifier(), promise = WTFMove(promise)] (auto& context) mutable { 
+            ASSERT(isMainThread());
 
-        context->postTask([parameterDescriptor, promise = WTFMove(promise),  permissionState = WTFMove(permissionState)](auto& context) mutable {
-            if (!permissionState) {
-                promise.reject(Exception { NotSupportedError });
+            auto* document = downcast<Document>(&context);
+            if (!document->page()) {
+                promise.reject(Exception { InvalidStateError, "The page does not exist"_s });
                 return;
             }
-            promise.resolve(PermissionStatus::create(context, *permissionState, parameterDescriptor));
-        });
+
+            PermissionController::shared().query(ClientOrigin { context.topOrigin().data(), WTFMove(originData) }, PermissionDescriptor { permissionDescriptor }, *document->page(), [context = Ref { context }, contextIdentifier, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
+                if (!context.get().globalObject()) {
+                    promise.reject(Exception { InvalidStateError, "The context is invalid"_s });
+                    return;
+                }
+
+                ScriptExecutionContext::postTaskTo(contextIdentifier, [promise = WTFMove(promise), permissionState, permissionDescriptor](auto& context) mutable {
+                    if (!permissionState)
+                        promise.reject(Exception { NotSupportedError, "Permissions::query does not support this API"_s });
+                    else
+                        promise.resolve(PermissionStatus::create(context, *permissionState, permissionDescriptor));
+                });
+            });
+        };
+
+        workerGlobalScope->thread().workerLoaderProxy().postTaskToLoader(WTFMove(completionHandler));
+        return;
+    }
+
+    PermissionController::shared().query(ClientOrigin { context->topOrigin().data(), WTFMove(originData) }, PermissionDescriptor { permissionDescriptor }, *document->page(), [this, protectedThis = Ref { *this }, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
```

Comment:
Why are we capturing `this` and `protectedThis`, seems we could just capture `document = Ref { document }`.

Reply:
Fixed!